gem/utils.py

```diff
@@ -88,3 +94,51 @@ def __exit__(self, exc_type, exc_value, traceback):
         assert self.state is variable._head
         value, variable._head = variable._head
         self.state = None
+
+
+@functools.singledispatch
+def safe_repr(obj: Any) -> str:
+    """Return a 'safe' repr for an object, accounting for floating point error.
+
+    Parameters
+    ----------
+    obj :
+        The object to produce a repr for.
+
+    Returns
+    -------
+    str :
+        A repr for the object.
+
+    """
+    raise TypeError(f"Cannot provide a safe repr for {type(obj).__name__}")
+
+
+@safe_repr.register(str)
+def _(text: str) -> str:
+    return text
+
+
+@safe_repr.register(numbers.Integral)
+def _(num: numbers.Integral) -> str:
+    return repr(num)
+
+
+@safe_repr.register(numbers.Real)
+def _(num: numbers.Real) -> str:
+    return format_float(num)
```

Comment:
I think we can still break our cache in parallel:
```
>>> format_float(1.0000000000000006)  # value on rank 0 (say)
'1.0000000000000007'
>>> format_float(1.0000000000000005)  # value on rank 1
'1.0000000000000004'
```
We might need to set correct precision in some manner.

Reply:
I doubt that would be sufficient, unfortunately. That means that we regard `1.+1.e-9` and `1.+2.e-9` as the same as far as quadrature points (+ caching) are concerned. I do not think we want to do that behind users.

Reply:
done

Reply:
What do you suggest Koki?

Reply:
I actually don't know what the best thing to do here is. I think, for double, if we pick 16 sig fig, it will almost certainly cause disagreement across ranks. I might pick 14 sig fig first. The reasoning is that users will not complain even if we regard `1.+1.e-14` and `1.+2.e-14` as the same, and using the leading 14 will likely to wipe out round off error. I could be that it is too strict, but `pyop2.parallel_cache` can check that all ranks get the same cache key and can conservatively raise exception otherwise. If we observe that the exception is raised, we can think about relaxing (to 13 digits, say).

Reply:
Amended

Reply:
Looks good to me. Could you also include a check in `pyop2.parallel_cache` so that it would explicitly raise exception if computed keys are inconsistent across ranks?

Reply:
That's already there: https://github.com/firedrakeproject/firedrake/pull/3989/files#diff-6f3c035d9e46a57067f2cfe0ce6269c58a488a7b0bebf0be005c4f9c37a55fbcR481

Reply:
Good.